SharedUtils/src/main/java/com/braintreepayments/api/SignatureVerification.java

```diff
@@ -47,38 +43,28 @@ static boolean isSignatureValid(Context context, String packageName,
             return false;
         }
 
-        InputStream certStream = null;
-        boolean validated = (signatures.length != 0);
-        for (Signature signature : signatures) {
-            try {
-                certStream = new ByteArrayInputStream(signature.toByteArray());
-
-                X509Certificate x509Cert =
-                        (X509Certificate) CertificateFactory.getInstance("X509")
-                                .generateCertificate(certStream);
-
-                String subject = x509Cert.getSubjectX500Principal().getName();
-                String issuer = x509Cert.getIssuerX500Principal().getName();
-                int actualPublicKeyHashCode = x509Cert.getPublicKey().hashCode();
+        if (signatures.length == 0) {
+            return false;
+        }
 
-                validated &= (certificateSubject.equals(subject) &&
-                        certificateIssuer.equals(issuer) &&
-                        publicKeyHashCode == actualPublicKeyHashCode);
+        for (Signature signature : signatures) {
+            if (signature != null) {
+                String currentSignature;
+                try {
+                    MessageDigest md = MessageDigest.getInstance("SHA-256");
+                    md.update(signature.toByteArray());
+                    currentSignature = Base64.encodeToString(md.digest(), Base64.DEFAULT);
+                } catch (NoSuchAlgorithmException e) {
+                    return false;
+                }
 
+                boolean validated = base64EncodedSignature.equals(currentSignature);
                 if (!validated) {
                     return false;
```

Comment:
The legacy approach for signature verification was to return false if there are any signatures that don't match (i.e. the app as more than one signature). Does that seem correct still?

Reply:
Sounds right to me